crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -267,25 +294,28 @@ impl Paginator {
             });
         }
 
+        let reset_state_guard =
+            ResetStateGuard { target: PaginatorState::Idle, state: self.state.clone() };
+
         let mut options = MessagesOptions::new(dir).from(token.as_deref());
         options.limit = num_events;
 
-        let response_result = self.room.messages(options).await;
+        // In case of error, the state is reset to idle automatically thanks to
+        // reset_state_guard.
+        let response = self.room.messages(options).await?;
 
-        // In case of error, reset the state to idle.
-        let response = match response_result {
-            Ok(res) => res,
-            Err(err) => {
-                self.state.set(PaginatorState::Idle);
-                return Err(err);
-            }
-        };
+        // NOTE: it's super important to not have any `await` after this point, since we
+        // don't want the task to be interrupted anymore, or the internal state
+        // may be incorrect.
 
         let hit_end_of_timeline = response.end.is_none();
-        *token_lock.lock().await = response.end;
+        *token_lock.lock().unwrap() = response.end;
 
         // TODO: what to do with state events?
 
+        // Forget the reset state guard, so its Drop method is not called.
```

Comment:
It sounds like you'd want to use `ManuallyDrop` instead? I'm not sure, but it's not out of the question that you have some heap allocated data inside the guard and `forget` will not deallocate those:

> Takes ownership and “forgets” about the value without running its destructor.
> 
> Any resources the value manages, such as heap memory or a file handle, will linger forever in an unreachable state. However, it does not guarantee that pointers to this memory will remain valid.